app/controllers/tweets_controller.rb

```diff
@@ -0,0 +1,90 @@
+class TweetsController < ApplicationController
+
+  # CREATE
+
+  get "/tweets/new" do # (must validate session ID)
+    @flash_message = session[:flash] if session[:flash]
+    session[:flash] = nil
+    if UserHelper.logged_in?(session)
+      @flash_message = session[:flash] if session[:flash]
+      session[:flash] = nil
+      erb :'/tweets/new'
+    else
+      redirect "/login"
+    end
+  end
+
+  post "/tweets" do
+    @tweet = Tweet.new(params)
+    @tweet.user = UserHelper.current_user(session)
+    if @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      session[:flash] = "Cannot post empty tweet"
+      redirect "/tweets/new"
+    end
+  end
+
+  # READ
+
+  get "/tweets" do
+    if @current_user = UserHelper.current_user(session)
+      @tweets = Tweet.all
+      erb :'tweets/index'
+    else
+      redirect "/login"
+    end
+  end
+
+  get "/tweets/:id" do
+    if @current_user = UserHelper.current_user(session)
+      @tweet = Tweet.find(params[:id])
+      erb :'/tweets/show'
+    else
+      redirect "/login"
+    end
+  end
+
+  # UPDATE
+
+  get "/tweets/:id/edit" do # (must validate session ID)
+    @flash_message = session[:flash] if session[:flash]
+    session[:flash] = nil
+    @current_user = UserHelper.current_user(session)
+    if !@current_user
+      redirect "/login"
+    else
+      @tweet = Tweet.find(params[:id])
+      if @tweet.user == @current_user
+        erb :'/tweets/edit'
+      else
+        redirect "/tweets"
+      end
+    end
+
+  end
+
+  patch "/tweets/:id" do
+    @tweet = Tweet.find(params[:id])
+    if @tweet.update(params[:tweet])
+      redirect "/tweets/#{@tweet.id}"
+    else
+      session[:flash] = "Cannot post empty tweet"
+      redirect "/tweets/#{@tweet.id}/edit"
+    end
+  end
+
+  # DELETE
+
+  delete "/tweets/:id/delete" do # (must validate session ID)
+    @current_user = UserHelper.current_user(session)
+    @tweet = Tweet.find(params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.